<!DOCTYPE html>
<html>
    <head>
        <title> Project documentation</title>
    </head>
    <body>
        <h2> Description of the project</h2>
        <p> The aim of the project is to adress some questions on the human genome by analyzing and performing operations on a GFF3 file. 
            The input for this software is in fact a GFF3 file, with a size of about 37 MB,
            as it contains only the annotation of the sequences. A GFF3 file is organized in 9 
            different columns, which include: </p>
        <ol> 
            <li><i><b>'seqid'</b></i>: the unique ID of the feature</li>
			<li><b><i>'source'</i></b>: the software or database name where the feature derives from </li>
			<li><b><i>'type'</i></b>: the type of the feature</li>
			<li><b><i>'start'</i></b>: the start position of the feature</li>
			<li><b><i>'end'</i></b>: the end position of the feature</li>
			<li><b><i>'score'</i></b>: the score of the feature </li>
			<li><b><i>'strand'</i></b>: the strand of the feature (+, -) </li>
			<li><b><i>'phase'</i></b>: represents the reading fram for CDS features</li>
			<li><b><i>'attributes'</i></b>: a list separated by semicolomns of tag-value pairs providing additional information</li>
        </ol>
        <p>The following documentation shows how we implemented the
            software and we will go through  <a href="#software_analysis">Software Analysis</a> and <a href="#software_design">Software Design</a>  </p>
        
        <a id="software_analysis">
            <h2>Software Analsyis  </h2>
        </a>
            <p> CRC cards for 4 classes: </p>
            <h4> Dataset Reader</h4>
            <img src="{{ url_for('static', filename='class3.png') }}" class="img-fluid">
            <h4> GFF3 Dataset Reader</h4>
            <img src="{{ url_for('static', filename='class4.png') }}" class="img-fluid">
            <h4> Dataset</h4>
            <img src="{{ url_for('static', filename='class1.png') }}" class="img-fluid">
            <h4>GFF3 Dataset</h4>
            <img src="{{ url_for('static', filename='class2.png') }}" class="img-fluid">
        
        <a id="software_design">
            <h2> Software design </h2>
        </a>
        <p> To represent the structure of our software, we also used an UML class diagram. In fact, the class diagram shows the relationships and interactions between the classes of the software.</p>
        <figure>
			<img src="/static/UML.png" alt="UML class diagram">
		</figure>
        <p> The UML class diagram presented above provides a graphical representation of the four classes: Dataset, DatasetGFF3, DatasetReader, and DatasetReaderGFF3, along with their attributes and methods, and the relationships between them.
            <br>
            In general: 
            <li>
                Dataset is a general abstract interface that has no actual implementation.
                DatasetGFF3, on the other hand, is a concrete class that extends the Dataset class. 
                The relationship between Dataset and DatasetGFF3 is a realization.
            </li>
            <li>Similarly, DatasetReader is a general interface that has no implementation. 
                DatasetReaderGFF3 is a concrete class that extends the DatasetReader class. The relationship between DatasetReader and DatasetReaderGFF3 is also a realization.
            </li>
            <li>
                The relationship between DatasetGFF3 and DatasetReaderGFF3 is a dependency,in fact, DatasetGFF3 uses the 'read()' method of the DatasetReaderGFF3 class for its implementation. 
                Therefore, any changes made to the DatasetReaderGFF3 class can have an impact on the DatasetGFF3 class.
            </li>

            Furthermore, the UML diagram contains the name, attributes, and methods for each class, which is represented by a box. 
            The box helps to understand the input and output types of each method and attribute.
        </p>
        <p>
            Also, the software is organised into 5 python modules:
            <a href="#dataset_reader">dataset_reader</a>, 
            <a href="#dataset">dataset</a>, 
            <a href="#decorator">decorator</a>, 
            <a href="#operations">operations</a>,
            <a href="#app">app</a>
        
        </p>
        <a id="dataset_reader">
            <h2><b>dataset_reader</b></h2>
        </a>
        <p>
            The 'dataset_reader' module reads the GFF3 file, and it contains 2 classes:
            <ul>
                <li>The Reader class is a general abstract interface that defined the procedures that a dataset reader is provided with</li>
                <li>The GFF3Reader class, which is specific to reading GFF3 datasets, and which returns a dataset object as output that is a wrapper around a Pandas DataFrame. </li>
            </ul>
        </p>
        
        <a id="dataset">
            <h2><b>dataset</b></h2>
        </a>
        <p>
            The dataset module, as for the dataset_reader, distinguishes between a general dataset and a GFF3 data. In fact, it also contains 2 classes:
            <ul>
                <li>The Dataset class, which is generic</li>
                <li>The GFF3Dataset, which is specific for GFF3 files.</li>
            </ul>
        </p>

        <a id="decorator">
            <h2><b>Decorator</b></h2>
        </a>
        <p> The decorator module consists in a decorator used in order to manage active operations. In fact, only operations that are registered as active can be performed over the dataset object. 
            This module contains two functions: 
            <ul>
                <li>The active() function, which returns a dictionary containing boolean values that indicate if certain operations are active or not.</li>
                <li> The decorator function, which is a higher-order function that takes another function function as its argument, and returns a wrapper function. 
                    The wrapper function checks if the function is active or not, using the active() dictionary. 
                    If the operation is active, it calls the original function function with its arguments and returns its output. 
                    If the operation is not active, it returns a message indicating that the operation is not active.</li>
            </ul>
         </p>
        
        <a id="operations">
            <h2><b>Operations</b></h2>
        </a>
        <p>This module defines different functions, each of which perform an operation over a pandas dataframe, 
            in fact, the functions use pandas method to manipulate data. 
            Each function is decorated with @decorator, to check whether they are active before their execution.
            Also, the 'active' function returns a dictionary in which each operation is mapped to a boolean value, indicating whether or not it is active. 
            The operations provided by our software are:
            <ul>
                <li> 'get_basic_info', which returns the name and data type of each column of the dataframe.</li>
                <li> 'get_unique_seqids', which returns a list of the unique sequence IDs available in the dataframe. </li>
                <li> 'get_unique_operations', which returns a list of unique types of operations available in the dataframe</li>
                <li>'count_features_by_source', which counts the number of features provided by the same source.</li>
                <li>'count_entries_by_type', which counts the number of entries for each type of operation.</li>
                <li>'entire_chromosome', which derives a new dataframe containing only the information about entire chromosomes.</li>
                <li>'fraction_unassembled', which calculates the fraction of unassembled sequences from source GRCh38</li>
                <li>'get_selected_sources', which obtains a new dataset containing only entries from source ensembl , havana and
                    ensembl_havana. </li>
                <li>'count_selected_operations', which counts the number of entries for each type of operation for the dataset containing containing only entries from source ensembl , havana and ensembl_havana.</li>
                <li>'get_gene_names', which returns the gene names from the dataset containing containing only entries from source ensembl ,
                    havana and ensembl_havana .</li>
            </ul>
        </p>

        <a id="app">
            <h2><b>App</b></h2>
        </a>
        <p>This module consists in the implementation of a Flask web application that provides an user interface to perform operations on a GFF3 dataset.
            <ul>
                <li>First of all, the GFF3 dataset is converted into a Pandas DataFrame thanks to the 'get_df()' function.</li>
                <li>Then a dictionary called 'av_operations' is created, which maps the name of an operation to its description.
                    And also, two empty dictionaries for 'active_operations' and 'active_operations_descr' are created.
                </li>
                <li>For each operation in av_operations, it is added to active_operations and its description to active_operations_descr</li>
                <li> Thenk the Flask routes for the homepage, active_operations, execute_operation and project_document are 
                    created thanks to the render_template() function. 
                </li>
            </ul>
        </p>

        <h3> Templates</h3>
        <p> In order to run the Flask application implemented in app.py, we implemented 4 routes in app.py, each of which is connected to an HTML file.
            In fact, we also developed 4 HTML files:
            <ul>
                <li>Homepage.html : it is the main page which is presented when a user accesses the application. It shows the other available pages provided by the application.  </li>
                <li>Active_operations.html: it shows the list of active operations, from which the user can select an operation to perform over the dataset. 
                    Once an operation is selected, the user is directed to the execute_operation page. </li>
                <li>operation_result.html: it prints the output for the operation selected in active_operations.</li>
                <li>Project_document.html: Shows the implementation of this project, by its descriptiion in terms of software analysis and software design. </li>
            </ul>
        </p>


    </body>
</html>


